common/formatting/token_partition_tree.cc

```diff
@@ -181,4 +182,236 @@ TokenPartitionTree* MoveLastLeafIntoPreviousSibling(TokenPartitionTree* tree) {
   return rightmost_leaf_parent;
 }
 
+//
+// TokenPartitionTree class wrapper used by AppendFittingSubpartitions and
+// ReshapeFittingSubpartitions for partition reshaping purposes.
+//
+class TokenPartitionTreeWrapper {
+ public:
+  TokenPartitionTreeWrapper(const TokenPartitionTree& node_) : node(&node_) {}
+
+  // Grouping node with no corresponding TokenPartitionTree node
+  TokenPartitionTreeWrapper(const UnwrappedLine& uwline_) : node(nullptr) {
+    unwrapped_line = std::make_shared<UnwrappedLine>(uwline_);
+  }
+
+  TokenPartitionTreeWrapper() = delete;
+
+  // Return wrapped node value or concatenation of subnodes values
+  const UnwrappedLine& Value() const {
+    if (node) {
+      return node->Value();
+    } else {
+      return *unwrapped_line;
+    }
+  }
+
+  // Concatenate subnodes value with other node value
+  UnwrappedLine Value(const TokenPartitionTree& other) const {
+    CHECK((node == nullptr) && (unwrapped_line != nullptr));
+    UnwrappedLine uw = *unwrapped_line;
+    uw.SpanUpToToken(other.Value().TokensRange().end());
+    return uw;
+  }
+
+  // Update concatenated value of subnodes
+  void Update(const VectorTree<TokenPartitionTreeWrapper>* child) {
+    const auto& token_partition_tree_wrapper = child->Value();
+    const auto& uwline = token_partition_tree_wrapper.Value();
+
+    unwrapped_line->SpanUpToToken(uwline.TokensRange().end());
+  }
+
+  // Fix concatenated value indentation
+  void SetIndentationSpaces(int indent) {
+    CHECK((node == nullptr) && (unwrapped_line != nullptr));
+    unwrapped_line->SetIndentationSpaces(indent);
+  }
+
+  // Return address to wrapped node
+  const TokenPartitionTree* Node() const {
+    return node;
+  }
+
+ private:
+  // Wrapped node
+  const TokenPartitionTree* node;
+
+  // Concatenated value of subnodes
+  std::shared_ptr<UnwrappedLine> unwrapped_line;
```

Comment:
`unwrapped_line_`

Can this be a `std::unique_ptr`?  If not, explain in comment.

Reply:
Changed to `std::unique_ptr`